### DISCUSSION TITLE
Is it possible to expose Usage/Monitoring/Billing/Cost data via the API?

### DISCUSSION THREAD

**Discussion by danieljmac** [vercel/community#2063] [Idea]
### Summary

Hey folks, new here and in the process of moving a bunch of client projects over to NextJS + Vercel. Currently doing some R&D for a particular client requirement, and hoping to get some insight on if/how something would be possible?

The TLDR is I'm basically trying to find a way to make the usage+cost+billing data you can visualize through the Usage tab or Monitoring queries, programmatically accessible so that I can auto generate weekly cost and usage reports into a 3rd party dashboard, as well as exported to csv.

Is this possible? I see that with the Monitoring addon subscription, you can generate custom queries that would fetch the data I need, and I also see that you can manually click to export a query to json or csv... but is there a way automate this on a schedule? Doesn't look like billing/usage data is available via API, but is there a way to setup a cron to run a Monitoring Query and maybe run a CLI command to automate the json/csv export?

Thanks a bunch, and excited to hopefully be more involved here in the forums!

**Comment by amyegan** [comment]
Thanks for reaching out, @danieljmac. There is not an way to automate this yet, but I'll move this to the "Ideas" category and share it with the team

**Reply by danieljmac** [reply]
Thanks for the quick reply, Amy! Just curious, would it be against any Vercel policy/rules to do something like run a cypress automation on a cron, to actually login to my own dashboard during a test, and intercept + extract the various usage datapoints there directly from my rendered Vercel Dashboard pages?

**Comment by ucheNkadiCode** [comment]
@danieljmac hey! Could you explain the ultimate output you're looking for? 

I'd like to understand if this flow is what you're looking for 

- You write a monitoring query and receive bandwidth by request path
- You click a NEW button called "Copy URL" 
- You're now able to paste that URL (with an auth token etc) in any browser and receive a JSON payload of those same results you see.

OR are you more looking for

- Navigate to the Usage Tab
- On any of the charts you can click "Copy URL" 
- And once again you're able to paste that URL in a browser and receive a JSON payload of the same results for that timespan?

Then of course with that URL you could call it as you please in external systems

**Reply by danieljmac** [reply]
Thanks for the response! I think I could achieve what I'm after with either of those approaches. 

Ultimately, what I'm trying to do is just get a json breakdown of accurate cost accrual data for a given billing period, so I can schedule a job to automatically fetch the current usage cost breakdown at least once per day. 

**The actual real-world use case I need it for:**
I manage the full tech stack for some of my bigger clients. Monthly costs are flying in from all angles (SendGrid, DBs, Cloud/Servers/Hosting, Marketing/Email Platforms, Plugin subs, etc). It was a nightmare trying to tell their finance depts to just login to each platform account every week to assess financials. So part of the service I provide them is a custom client dashboard I built, that aggregates current usage costs from all active providers, and rolls everything up on a single Costs page, so they can login whenever to see the breakdown of their total tech usage costs for the current billing period. 

Most of these are currently running on GCP, so it's all hooked up to pull the network + compute instance costs from Google. But I'm migrating a lot of platform stuff to NextJS and also want to make the move from GCP to Vercel, but just need to figure out a way to keep the cloud platform portion of this live cost reporting still connected in their client dashboards.

**Reply by ucheNkadiCode** [reply]
Absolutely amazing and clear breakdown. Thank you! Our team now understands the ask and will need to circulate this internally. We appreciate you letting us know

**Reply by chrillo** [reply]
I would like to second what danieljmac said and maybe add one more usecase. We currently got burrned very badly by a GCP bill for a DDOS attack and having access to the monitoring data would allow us to pull these metrics into our grafana & alerting infrastructure. (vercel very recently added cost alerts via webhooks which help a lot in that regard, it would still be great to have programmatic access to the data)

**Reply by mitjafelicijan** [reply]
Is there any information about adding this API endpoint? It's been almost a year and nothing but radio silence.

Based on the latest news about high bills accumulated on Vercel, this should be a priority to avoid bad press if nothing else (this way you could say there are API's you could hook in).

**Reply by leerob** [reply]
You can use the webhook functionality of Spend Management to get alerts as your spend changes for metered usage: https://vercel.com/blog/introducing-spend-management-realtime-usage-alerts-sms-notifications

Open to suggestions for how to improve this spend management functionality, however, we are not planning on a standalone API at this time. The next change we're making here will be for automatic pausing on the spend amount reached.

**Reply by mitjafelicijan** [reply]
That's great news. Do you maybe have a timeline when autopausing will be available?

As for the standalone API, that is a shame. Having programmatic access to information like this is becoming increasingly needed these days to build internal dashboards and measure costs, since the web is more and more fragmented and people use different services and keeping up with the costs has become a full-time job.
